### PR TITLE
First draft of custom event handler support

### PR DESCRIPTION
**Pull request by etburke**
Addresses https://github.com/samuelint/langchain-openai-api-bridge/issues/41

Apologies, I'm a Python novice. Please don't hesitate to reject this approach and/or provide feedback/guidance.

### CODE REVIEW

File: tests/test_functional/fastapi_chat_completion_openai/server_openai.py
Comment (samuelint) [inline tests/test_functional/fastapi_chat_completion_openai/server_openai.py:1]: It would be preferable to have a new test instead of modifying the existing one. We still want to test the use case without the `lambda`.

File: langchain_openai_api_bridge/core/utils/pydantic_async_iterator.py [hunk 1]
```diff
@@ -5,4 +5,4 @@
 
 async def ato_dict(async_iter: AsyncIterator[BaseModel]) -> AsyncIterator[dict]:
     async for obj in async_iter:
-        yield obj.dict()
```
Comment (samuelint) [inline langchain_openai_api_bridge/core/utils/pydantic_async_iterator.py:8]: Since langchain use pydantic V1 by default. `.dict` should be kept until `langchain` use V2 by default.

File: tests/test_unit/chat_completion/test_langchain_stream_adapter.py [hunk 2]
```diff
@@ -43,3 +43,35 @@ async def test_stream_contains_every_on_chat_model_stream(
...
+
+    
+    @pytest.mark.asyncio
+    @patch(
+        "langchain_openai_api_bridge.chat_completion.langchain_stream_adapter.to_openai_chat_completion_chunk_object",
+        side_effect=lambda event, id, model, system_fingerprint: (
+            ChatCompletionChunkStub({"key": event["data"]["chunk"].content})
+        ),
+    )
+    async def test_stream_contains_every_custom_handled_stream(
```
Comment (samuelint) [inline tests/test_unit/chat_completion/test_langchain_stream_adapter.py:55]: Nice test!

File: langchain_openai_api_bridge/core/base_agent_factory.py [hunk 3]
```diff
@@ -4,7 +4,11 @@
 
 
 class BaseAgentFactory(ABC):
-
+    
+    @classmethod
+    def custom_event_handler(self, event):
```
Comment (samuelint) [inline langchain_openai_api_bridge/core/base_agent_factory.py:9]: I do not think having a `custom event handler`, is related to building an agent. I think event handler should be passed as a kwarg of `LangchainOpenaiApiBridgeFastAPI`. or `.bind_openai_chat_completion`.

You can see an example of similar implementation here: https://github.com/samuelint/langchain-openai-api-bridge/blob/dc398f0deab77ff50e18a115f6a8913efa34fa5c/tests/test_functional/fastapi_assistant_agent_openai/assistant_server_openai.py#L43-L48

**Review by samuelint — CHANGES_REQUESTED** [review]
Sorry for the delay, I just came back from vacation :)

File: langchain_openai_api_bridge/core/base_agent_factory.py (hunk 3 above)
Comment (etburke) [reply langchain_openai_api_bridge/core/base_agent_factory.py:9]: I actually implemented it as a kwarg of `LangchainOpenaiApiBridgeFastAPI` before talking myself into refactoring it. I'll move it to a kwarg of `.bind_openai_chat_completion` as you've already established that as a pattern.

File: tests/test_functional/fastapi_chat_completion_openai/server_openai.py
Comment (etburke) [reply tests/test_functional/fastapi_chat_completion_openai/server_openai.py:1]: Sorry, I meant to go back and split that into two tests. Will do.

File: langchain_openai_api_bridge/fastapi/chat_completion_router.py [hunk 4]
```diff
@@ -16,10 +16,11 @@
 
 def create_chat_completion_router(
     tiny_di_container: TinyDIContainer,
+    custom_event_handler: callable = lambda event: None,
 ):
-    chat_completion_router = APIRouter(prefix="/chat/completions")
+    chat_completion_router = APIRouter(prefix="/chat")
 
-    @chat_completion_router.post("/")
+    @chat_completion_router.post("/completions")
```
Comment (etburke) [inline langchain_openai_api_bridge/fastapi/chat_completion_router.py:23]: I meant to mention I included this unrelated change. In my case I had to include the following for the Vercel SDK to successfully hit my custom endpoint:
```
# Remove trailing slashes from routes, because LangchainOpenaiApiBridgeFastAPI adds them
for route in app.routes:
    if route.path.endswith("/"):
        app.routes.remove(route)
        app.add_api_route(route.path[:-1], route.endpoint, methods=route.methods)
```
This seems to resolve that without causing side effects.

**Comment by etburke** [comment]
Okay! I've made some updates attempting to address your initial feedback. Let me know if you have any other thoughts.

**Comment by etburke** [comment]
@samuelint any thoughts on this?

File: langchain_openai_api_bridge/fastapi/chat_completion_router.py (hunk 4 above)
Comment (samuelint) [reply langchain_openai_api_bridge/fastapi/chat_completion_router.py:23]: Make sure Integration Tests are all passing. If this is the case, everything should be fine.

File: langchain_openai_api_bridge/chat_completion/langchain_stream_adapter.py [hunk 5]
```diff
@@ -22,20 +21,29 @@ async def ato_chat_completion_chunk_stream(
...
+            else:
+                kind = event["event"]
+                match kind:
+                    case "on_chat_model_stream":
+                        yield to_openai_chat_completion_chunk_object(
+                            event=event,
+                            id=id,
+                            model=self.llm_model,
+                            system_fingerprint=self.system_fingerprint,
+                        )
```
Comment (samuelint) [inline langchain_openai_api_bridge/chat_completion/langchain_stream_adapter.py:46]: Those 2 functions do not need to be repeated. This is violating the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). 

`event` should be in a variable on top  instead of re-writing the entire `yield to_openai_chat_completion_chunk_object( ...`

File: langchain_openai_api_bridge/chat_completion/langchain_stream_adapter.py [hunk 6]
```diff
@@ -22,20 +21,29 @@ async def ato_chat_completion_chunk_stream(
         self,
         astream_event: AsyncIterator[StreamEvent],
         id: str = "",
+        custom_event_handler = lambda event: None,
```
Comment (samuelint) [inline langchain_openai_api_bridge/chat_completion/langchain_stream_adapter.py:24]: I think the variable should be `event_adapter`. Since it's adapting the original event. This comment applies for the entire code base change.

File: tests/test_functional/fastapi_chat_completion_openai/server_openai.py [hunk 7]
```diff
@@ -36,9 +35,16 @@ def create_agent(dto: CreateAgentDto):
         api_key=dto.api_key,
     )
 
-
 bridge = LangchainOpenaiApiBridgeFastAPI(app=app, agent_factory_provider=create_agent)
 bridge.bind_openai_chat_completion(prefix="/my-custom-path")
 
+def custom_event_handler(event):
```
Comment (samuelint) [inline tests/test_functional/fastapi_chat_completion_openai/server_openai.py:41]: Existing tests should not be changed. 
Reason: We do not want to alter the use case of not having a "custom event handler".  We still need to test that the library works without it.

A new integration test should be made for this specific use case.

**Comment by samuelint** [comment]
> @samuelint any thoughts on this?

I've reviewed again, sorry for the delay.
After that everything should be fine. Let me know if you have any questions.

**Comment by etburke** [comment]
Alright, changes made. All tests passing.

File: tests/test_functional/fastapi_chat_completion_openai/server_openai.py [hunk 8]
```diff
@@ -36,9 +35,16 @@ def create_agent(dto: CreateAgentDto):
...
 bridge.bind_openai_chat_completion(prefix="/my-custom-path")
 
+def event_adapter(event):
+    kind = event["event"]
+    match kind:
+        case "on_chat_model_stream":
+            return event
+
+bridge.bind_openai_chat_completion(prefix="/my-custom-events-path", event_adapter=event_adapter)
+
```
Comment (samuelint) [inline tests/test_functional/fastapi_chat_completion_openai/server_openai.py:48]: This seem to be dead code. Ideally the whole file should not have changes

**Review by samuelint — APPROVED** [review]

File: tests/test_functional/fastapi_chat_completion_openai/server_openai.py (hunk 8 above)
Comment (etburke) [reply tests/test_functional/fastapi_chat_completion_openai/server_openai.py:48]: Oh, yep. That's dead code. I'll remove that.

**Comment by etburke** [comment]
Thanks @samuelint !
